text/3334-niche.md

```diff
@@ -0,0 +1,480 @@
+the range of *valid* values. Different types may find one or the other case
+simpler. This RFC chooses to define the range of *invalid* values for three
+reasons:
+- As an arbitrary choice, because we need to pick one or the other (see above).
+- The most common case will be a single invalid value, for which defining
+  invalid values results in simpler code.
+- This mechanism commonly goes by the name `niche`, and `niche` also refers to
+  the invalid value. So, an attribute defining the niche of a type most
+  naturally refers to the invalid value.
+
+Note that the compiler already supports having a niche in the middle of a
+type's possible values; internally, the compiler represents this by defining a
+valid range that wraps around the type's possible values. For instance,
+`#[niche(value = 42)]` gets represented internally in the compiler as a valid
+range starting at 43 and ending at 41.
+
+We could define *only* single-value niches, not ranges. However, the compiler
+already supports ranges internally, and the standard library already makes use
+of multi-value ranges, so this seems like an artificial limitation.
+
+We could define only ranges, not single-value niches, and users could express
+single-value niches via ranges, such as `0..=0`. However, that makes
+single-value niches more verbose to define, and makes mistakes such as `0..0`
+more likely. (This RFC suggests a lint to catch such cases, but the syntax
+should still attempt to guide users away from that mistake.)
+
+We could guarantee more usage of niches than just a single value; however, this
+would constrain the compiler in areas that still see active development.
+
+We could avoid guaranteeing the use of a single-value niche for `Option`;
+however, this would eliminate one of the primary user goals for such niches.
+
+We could require types to opt into the guaranteed use of a niche, separately
+from declaring a niche. This seems unnecessarily verbose, as well as limiting:
+we can't yet provide a full guarantee of all *future* uses we may want to
+guarantee, only of the limited single-value uses.
+
+We could implement niches using a lang-item type that uses const generics (e.g.
+`Niche<T, const RANGE: std::ops::Range<T>>`. This type would be useful
+regardless, and we should likely provide it if we can. However, this RFC
+advocates (eventually) building such a type on an underlying language-level
+building block like `niche`, and providing the underlying building blocks to
+the ecosystem as well.
+
+We could implement niches using a trait `Niche` implemented for a type, with
+associated consts for niche values. If we chose to do this in the future, the
+`#[niche(...)]` attribute could become forward-compatible with this, by
+generating the trait impl.
+
+We could use a syntax based on patterns, such as `struct S(u8 is 0..=32);` or
+`struct S(MyEnum is MyEnum::A | MyEnum::B)`. The `niche` attribute could be
+forward-compatible with this, by generating the appropriate patterns.
```

Comment:
I was not making a syntax nitpick, I was meaning the ideas around extending the type system with the notion of "types whose values all match a given pattern" i.e. `u8 is 0..10` being a standalone type. not a special case in type syntax for when in a field. Properly doing niches via the type system would not require any of the unsafe this RFC requires